modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py

```diff
@@ -273,14 +277,19 @@ def apply(
         list
             A list of `PandasOnRayDataframeVirtualPartition` objects.
         """
-        # If this is not a full axis partition, it already contains a subset of
-        # the full axis, so we shouldn't split the result further.
         if not self.full_axis:
+            # If this is not a full axis partition, it already contains a subset of
+            # the full axis, so we shouldn't split the result further.
             num_splits = 1
         if len(self.call_queue) > 0:
             self.drain_call_queue()
+        kwargs["args"] = deserialize(args)
```

Comment:
Do we need this `deserialize`? I thought these args would not be serialized. If we do need to deserialize, why?

Reply:
This is needed. One codepath where this happens (and is a root source of the issue) is from `broadcast_apply`. https://github.com/modin-project/modin/blob/60518ca737bf13cc37c8f1269985dcbb68dad4c7/modin/core/dataframe/pandas/partitioning/partition_manager.py#L372 Here, the constructor `cls.axis_partition()` can create `rt_axis_parts` that can contain type `PandasOnRayDataframeVirtualPartition` 

and the `list_of_blocks` property in that same function https://github.com/modin-project/modin/blob/60518ca737bf13cc37c8f1269985dcbb68dad4c7/modin/core/dataframe/pandas/partitioning/partition_manager.py#L379-L381

is a list of `oid`s, from https://github.com/modin-project/modin/blob/60518ca737bf13cc37c8f1269985dcbb68dad4c7/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py#L104-L119

I would even go as far as defensively deserializing `func` and `kwargs` too, to match the behavior below. What do you think? https://github.com/modin-project/modin/blob/e7cb2e82f8b9c7a68f82abdd3b6011d661230b7e/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py#L408-L410